src/timesyncwait/timesyncwaitd.c

```diff
@@ -0,0 +1,124 @@
+/*
+ * systemd service to wait until kernel realtime clock is synchronized
+ *
+ * Copyright 2018 Peter A. Bigot
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
+ * USA
+ */
+
+#include <errno.h>
+#include <signal.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/timerfd.h>
+#include <sys/timex.h>
+
+#include "sd-event.h"
+
+#include "missing.h"
+#include "time-util.h"
+
+typedef struct ClockState {
+        int fd;                        /* non-negative is descriptor from timerfd_create */
+        int adjtime_state;             /* return value from last adjtimex(2) call */
+        sd_event *event;               /* the systemd event loop object */
+        sd_event_source *event_source; /* non-null is the active io event source */
+} ClockState;
+
+static int io_handler(sd_event_source * s,
+                      int fd,
+                      uint32_t revents,
+                      void *userdata) {
+        ClockState *sp = userdata;
+        int rc = 0;
+        struct itimerspec its = {
+                .it_value.tv_sec = TIME_T_MAX,
+        };
+        struct timex tx = {0};
+
+        /* Release any event source and timer descriptor held from the last
+         * check, then set a timerfd to signal when the clock is set.  Do
+         * this before we check synchronized state to avoid a race
+         * condition. */
+        sp->event_source = sd_event_source_unref(sp->event_source);
+        if (0 <= sp->fd)
+                (void) close(sp->fd);
+        sp->fd = timerfd_create(CLOCK_REALTIME, TFD_NONBLOCK | TFD_CLOEXEC);
+        if (0 <= sp->fd)
+                rc = timerfd_settime(sp->fd, TFD_TIMER_ABSTIME | TFD_TIMER_CANCEL_ON_SET, &its, NULL);
```

Comment:
hmm, thinking about this, does this actually work in all cases? is adjtimex guaranteed to issue a TFD_TIMER_CANCEL_ON_SET event in all cases when it sets the synchronized bit? i mean, isn't there a logic in place that for small time changes the clock will be slowly adjusted and hence no CANCEL_ON_SET should be generated?

Reply:
I don't think it's guaranteed.  From what I can see:

* The kernel never marks the clock as synchronized, though it might mark it unsynchronized.
* Anything the kernel interprets as "setting the clock" will cancel the timer (including adjtimex sometimes, settimeofday always, and other operations sometimes).
* Small adjustments may not count as setting the clock, but they also may not change the synchronization state.  What happens depends on how they're applied.

So it is possible something might tell the kernel "you're synchronized" without doing something the kernel will interpret as "setting the clock", in which case the state change won't be signalled. 
 Nothing to be done about that except periodically poll, but I wouldn't bother until there's evidence somebody's (mis)using the API that way.  The existing approach works with timesyncd.

Reply:
ok, but this kinda makes me even more think that we shouldn't advertise this new tool as being generic and compatible with other NTP implementations, because it will only be compatible with well-behaving ones...

BTW, it would be good if you could add an explanation like the above as a comment to the relevant code, so that the next one looking at this will understand what is going on, and that this is safe with timesyncd.

Reply:
Next rev will add the comment and a Notes section in the man page noting the possible failures.